### PR TITLE
perf: Improve performance by creating temporary hard references in DependencyObjectStore

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): https://github.com/unoplatform/nventive-private/issues/102

## PR Type

What kind of change does this PR introduce?
- Refactoring

## What is the new behavior?

- Visual Tree traversal is improved by using parent weak reference materialization while UIElement instances are materialized
- Improvement in current applied theme selection
- Improve tree traversal algorithm by reducing type checking

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Directory.Build.targets [hunk 1]
```diff
@@ -20,11 +20,11 @@
...
 	<PackageReference Update="Uno.SourceGeneration" Version="3.0.0-dev.21" />
-	<PackageReference Update="Uno.Core" Version="2.0.0" />
-	<PackageReference Update="Uno.Core.Build" Version="2.0.0" />
+	<PackageReference Update="Uno.Core" Version="2.1.0" />
+	<PackageReference Update="Uno.Core.Build" Version="2.1.0" />
 	<PackageReference Update="Uno.Diagnostics.Eventing" Version="1.0.4" />
-	<PackageReference Update="Uno.Wasm.Bootstrap" Version="1.3.4" />
-	<PackageReference Update="Uno.Wasm.Bootstrap.DevServer" Version="1.3.4" />
+	<PackageReference Update="Uno.Wasm.Bootstrap" Version="2.0.0-dev.125" />
+	<PackageReference Update="Uno.Wasm.Bootstrap.DevServer" Version="2.0.0-dev.125" />
```
Comment (carldebilly) [inline src/Directory.Build.targets:27]: dev version?

File: src/Uno.UI/UI/Xaml/ApplicationHelper.cs [hunk 2]
```diff
@@ -30,6 +30,8 @@ public static string RequestedCustomTheme
 						Application.Current.RequestedTheme = ApplicationTheme.Light;
 					}
 				}
+
+				Application.UpdateRequestedThemesForResources();
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/ApplicationHelper.cs:34]: Is this related to perf?

File: src/Uno.UI/UI/Xaml/FrameworkElement.netstd.cs [hunk 3]
```diff
@@ -66,7 +66,7 @@ private protected sealed override void OnFwEltLoaded()
 		}
 
 		partial void OnLoadedPartial();
-		private protected virtual void OnLoaded() { }
+		private protected virtual void OnLoaded() {  }
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/FrameworkElement.netstd.cs:69]: Useful change ;-)

File: src/Uno.UI/UI/Xaml/ApplicationHelper.cs (hunk 2 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ApplicationHelper.cs:34]: Yes it is, it avoids the resources resolution to compute the active theme.

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.cs [hunk 4]
```diff
@@ -1657,6 +1679,40 @@ internal IEnumerable<ResourceBinding> GetResourceBindingsForProperty(DependencyP
...
 
+		/// <summary>
+		/// Enables hard references for internal fields for faster access
+		/// </summary>
+		/// <remarks>
+		/// Calling this method may cause memory leaks and must be used along with <see cref="DisableHardReferences"/>.
+		/// The use case is for FrameworkElement instances that can be Loaded and Unloaded, and for which the Unloaded
+		/// state is ensured to happen in such a way that memory leaks cannot happen.
+		/// </remarks>
+		internal void EnableHardReferences()
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.cs:1690]: Given that it depends on the feature flag, maybe `TryEnableHardReferences()`? (And if so, the same below)

File: src/Uno.UI/UI/Xaml/Application.cs [hunk 5]
```diff
@@ -91,6 +92,43 @@ public ApplicationTheme RequestedTheme
...
+					case ApplicationTheme.Light:
+						return "Light";
+					case ApplicationTheme.Dark:
+						return "Dark";
+					default:
+						throw new InvalidOperationException($"Theme {Application.Current.RequestedTheme} is not valid");
+				}
+			}
+
+			Application.Current.RequestedThemeForResources = Update();
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Application.cs:127]: Why do you create a nested method you invoke immediately? Sounds less performant ...

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.cs [hunk 6]
```diff
@@ -111,14 +113,16 @@ public static class TraceProvider
...
-			get => _parentRef?.Target;
+			get => _hardParentRef ?? _parentRef?.Target;
 			set
 			{
 				if (
 					!ReferenceEquals(_parentRef?.Target, value)
 					|| (_parentRef != null && value is null)
 				)
 				{
+					_hardParentRef = null;
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.cs:124]: What if `Parent` is set while hard references are enabled?

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 7]
```diff
@@ -452,33 +452,7 @@ public StaticResourceAliasRedirect(string resourceKey, XamlParseContext parseCon
...
-						case ApplicationTheme.Light:
-							return "Light";
-						case ApplicationTheme.Dark:
-							return "Dark";
-						default:
-							throw new InvalidOperationException($"Theme {Application.Current.RequestedTheme} is not valid");
-					}
-				}
-			}
+			public static string Active => Application.Current?.RequestedThemeForResources ?? "list";
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:455]:
```suggestion
			public static string Active => Application.Current?.RequestedThemeForResources ?? Default;
```

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs (hunk 7 above)
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:455]: "list" ???

```suggestion
			public static string Active => Application.Current?.RequestedThemeForResources ?? "Light";
```

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:455]: It was "Light" when the `Application.Current` is `null`

File: src/Directory.Build.targets (hunk 1 above)
Comment (jeromelaban) [reply src/Directory.Build.targets:27]: Not needed for this PR, will remove.

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.cs (hunk 6 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.cs]: In the current context, it does not change anything, because changing the parent generally implies the element has been unloaded and the references has been removed. Yet, it makes sense to call disable instead.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.cs]: > In the current context, it does not change anything, because changing the parent generally implies the element has been unloaded and the references has been removed. 

Not necessarily, though - in some cases `Parent` gets set manually, usually in cases where the 'visual parent' is a native view and we want the 'logical parent' to be a managed view.

I guess this not a big deal though? It should not just fall back on the weak ref without functional implications if I understand correctly, and it's sufficiently uncommon that it shouldn't have any significant perf impact.
